### PR TITLE
Add support new output format (fragmented-MP4 packaged with HLS)

### PR DESCRIPTION
**Pull request by emranemran**
This PR adds the option to generate fragmented-mp4s packaged with HLS. 

**TODOs:**
~~* Add option to generate different mp4 types in vod request handler~~
~~* Fix location of output dir for fmp4 in object-store~~
~~* Fix mp4 output response to Studio~~
~~-- Fix probing issue post-transcode~~
~~-- Fix fmp4 output to generate separate a/v tracks~~ --> will be a followup PR
~~-- Fix/add unit tests~~ (separate PR will be used for integration tests)
~~-- Some more testing in staging including regression testing~~

/Fix VID-341

### CODE REVIEW

File: handlers/upload.go
Comment (emranemran) [inline handlers/upload.go:85]: @leszko need your input on commit `3f9f5eb4013b2174ac76f088bba9e3788e6408eb`.  The `MP4` field in  `UploadVODRequestOutputLocation` is already set by studio when we need to generate an mp4 file. The change here adds an option to generate a regular old mp4 (MP4 set to `enabled`) or a fragmented mp4 (MP4 set to `fragmented`). In other words, we should only generate either a regular mp4 or fragmented-mp4. Do you see any issues or dependencies from Studio perspective?

Comment (thomshutt) [reply handlers/upload.go:85]: I don't love this - that parameter is for controlling _whether_ an MP4 is generated, rather than the type. Could we add the same parameter but for fMP4? That also opens the door to generating both, if that's ever something customers want

File: transcode/transcode.go
Comment (thomshutt) [inline transcode/transcode.go:219]: Drop before merge

File: transcode/transcode.go
Comment (thomshutt) [inline transcode/transcode.go:229]: Drop

File: transcode/transcode.go
Comment (thomshutt) [inline transcode/transcode.go:218]: Could we flip this so that we're not checking for a negative? Or like in my other comment, maybe support both?

**Comment by thomshutt** [comment]
Conceptually looks good, I like the splitting out of the two different MP4 generation methods rather than trying to combine them. Tag me once the tests are passing and it's ready for a final review.

File: handlers/upload.go
Comment (emranemran) [reply handlers/upload.go:85]: Will update.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:219]: Yeah, forgot to take out debug prints.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:218]: Yeah, will update this.

File: handlers/upload.go
Comment (emranemran) [reply handlers/upload.go]: Added new parameter for fMP4 which required a refactor - see latest commits (still wip but commit `f467915c364a0032847dec5f1445d2316570d29f` should be good for review).

**Comment by emranemran** [comment]
**Updates (still WIP)**
* refactored code to allow both regular MP4 and fragmented MP4 generation
* some debug prints need to be cleaned up still and tests need to be fixed/added. 
* if you are reviewing, then going commit by commit will be best as I tried to split them up logically for easier review.

**Review by leszko — COMMENTED** [review]
Added some early comments. I can review it again when you add all TODO point, but in general it looks good.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:187]: In case both output URLs for mp4 and fmp4 is malformed somehow in the incoming request. There are additional layers of checks elsewhere in catalyst-api and probably Studio as well. I can remove if you think it's redundant but thought we keep an extra layer of security here.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:282]: Not sure I understand. If mp4 output is detected, we will skip this ` if` block. When fmp4 output is requested, the files are `m4a` and ` m3u8` so I want to skip over those temporarily until the Studio integration is done.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:366]: `bool` here means that the given target url is valid so enable that type of output. I thought about using an enum here but I think that makes the logic a bit more convoluted here given the usage further down the code where we actually use these `bool`s.

**Comment by emranemran** [comment]
@leszko @thomshutt @mjh1 ready for one more round of review. 

Before merge: 
* run regression test suite

Post merge: 
* add integration tests for the transmux methods
* staging tests to check for regressions

**Review by leszko — COMMENTED** [review]
Added a few more comments, but in general it looks good. I played around with this and was able to generate the fmp4 files 🎉 

One general question, I had the following path: `/cakeout/hls/` set in my Storj bucket. Then, I've send the following request:
```
            "outputs": {
                "hls": "enabled",
                "mp4": "enabled",
                "fragmented_mp4": "enabled"
            }
```

Now, I was a little surprised by where the output files are located. Is that correct?
- HLS files are in `/cakeout/hls/index.m3u8`, `cakeout/hls/360p0/*`, `cakeout/hls/720p0/*`, `cakeout/hls/1080p0/*`=> ALL GOOD
- MP4 files are in `/cakeout/hls/360p0.mp4`, `/cakeout/hls/720p0.mp4`, `/cakeout/hls/1080p0.mp4` => ALL GOOD
- fMP4 files are in `/cakeout/hls/360p0/fmp4/360p0.m3u8`, `/cakeout/hls/360p0/fmp4/360p0.m4s` (and the corresponding dirs for `720p0` and `1080p0` => Is that correct? Don't we need some manifest files in the root dir like we have for HLS?

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:227]: Oops, forgot about all these.

File: transcode/transcode.go
Comment (emranemran) [reply transcode/transcode.go:225]: I'll clean this up. It's actually documenting the steps in MP4 generation if you scroll up a few comments.

File: handlers/upload.go
Comment (emranemran) [reply handlers/upload.go:28]: I found it useful to break it up here as some of the code becomes hard to distinguish between MP4 and FMP when you're reading through it.

File: video/transmux.go
Comment (emranemran) [reply video/transmux.go:29]: Yeah, i'll add this function as a test here as a follow-up PR after some staging tests: 
https://github.com/livepeer/catalyst-api/blob/cc8e4d396060494f1b7b6fee0a416be43f08b56f/test/features/vod.feature#L51

**Comment by emranemran** [comment]
> Added a few more comments, but in general it looks good. I played around with this and was able to generate the fmp4 files 🎉
> 
> One general question, I had the following path: `/cakeout/hls/` set in my Storj bucket. Then, I've send the following request:
> 
> ```
>             "outputs": {
>                 "hls": "enabled",
>                 "mp4": "enabled",
>                 "fragmented_mp4": "enabled"
>             }
> ```
> 
> Now, I was a little surprised by where the output files are located. Is that correct?
> 
> * HLS files are in `/cakeout/hls/index.m3u8`, `cakeout/hls/360p0/*`, `cakeout/hls/720p0/*`, `cakeout/hls/1080p0/*`=> ALL GOOD
> * MP4 files are in `/cakeout/hls/360p0.mp4`, `/cakeout/hls/720p0.mp4`, `/cakeout/hls/1080p0.mp4` => ALL GOOD
> * fMP4 files are in `/cakeout/hls/360p0/fmp4/360p0.m3u8`, `/cakeout/hls/360p0/fmp4/360p0.m4s` (and the corresponding dirs for `720p0` and `1080p0` => Is that correct? Don't we need some manifest files in the root dir like we have for HLS?

Yep, will fix this based on offline discussion with Rafal/Thom/Max

**Comment by leszko** [comment]
The current directory structure is like this:
image

So all the fmp4 files are embedded into the `fmp4` directory. I think it's ok, but may be a little confusing for the Transcode API users. Imagine you send the following request:
```
{
    "input": {
        "url": "some-video-url"
    },
    "storage": {
        "type": "s3",
        "endpoint": "https://gateway.storjshare.io",
        "credentials": {
            "accessKeyId": "access-key",
            "secretAccessKey": "secret-key"
        },
        "bucket": "my-bucket"
    },
    "outputs": {
        "hls": {
            "path": "/hls/"
        },
        "mp4": {
            "path": "/mp4/"
        },
        "fragmented_mp4": {
           "path": "/fmp4"
        }
    }
}
```

Now, you could probably expect that your fmp4 files will end it in the `/fmp4` location, but actually they'll end up in `/fmp4/fmp4` location.

I leave it up to you, but thinking about it again, I'd probably flatten the `fmp4` structure and then for our upload task, we would just add `/fmp4` in the request between task-runner<>catalyst.

**Review by leszko — APPROVED** [review]
LGTM 👍 

Added one comment again about the directory structure.

Plus, do we support (want to support) fmp4 in the mediaconvert pipeline?

File: transcode/transcode.go
Comment (mjh1) [inline transcode/transcode.go:358]: nit: Could you invert the if statement and return early to avoid the extra indentation?

File: transcode/transcode.go
Comment (mjh1) [inline transcode/transcode.go:329]: this needs to be moved down below the `err != nil` check

**Comment by mjh1** [comment]
> The current directory structure is like this: image
> 
> So all the fmp4 files are embedded into the `fmp4` directory. I think it's ok, but may be a little confusing for the Transcode API users. Imagine you send the following request:
> 
> ```
> {
>     "input": {
>         "url": "some-video-url"
>     },
>     "storage": {
>         "type": "s3",
>         "endpoint": "https://gateway.storjshare.io",
>         "credentials": {
>             "accessKeyId": "access-key",
>             "secretAccessKey": "secret-key"
>         },
>         "bucket": "my-bucket"
>     },
>     "outputs": {
>         "hls": {
>             "path": "/hls/"
>         },
>         "mp4": {
>             "path": "/mp4/"
>         },
>         "fragmented_mp4": {
>            "path": "/fmp4"
>         }
>     }
> }
> ```
> 
> Now, you could probably expect that your fmp4 files will end it in the `/fmp4` location, but actually they'll end up in `/fmp4/fmp4` location.
> 
> I leave it up to you, but thinking about it again, I'd probably flatten the `fmp4` structure and then for our upload task, we would just add `/fmp4` in the request between task-runner<>catalyst.

Well spotted 👍 annoyingly I think we'll have to add a check to add the prefix `/fmp4` only if the prefixes for hls and fmp4 aren't the same, as I don't think we'd want the flat structure of regular hls and fmp4 flat in the same folder.

**Comment by leszko** [comment]
> > Now, you could probably expect that your fmp4 files will end it in the `/fmp4` location, but actually they'll end up in `/fmp4/fmp4` location.
> > I leave it up to you, but thinking about it again, I'd probably flatten the `fmp4` structure and then for our upload task, we would just add `/fmp4` in the request between task-runner<>catalyst.
> 
> @leszko Well spotted 👍 annoyingly I think we'll have to add a check to add the prefix `/fmp4`if the prefixes for hls and fmp4 are the same, as I don't think we'd want the flat structure of regular hls and fmp4 in the same folder as they share a similar structure and might be confusing if they're mixed together.

Maybe, but for Transcode API we can leave it "up to the user" and for Upload Task just add the `/hls` prefix in task-runner.

**Comment by mjh1** [comment]
> Maybe, but for Transcode API we can leave it "up to the user" and for Upload Task just add the `/hls` prefix in task-runner.

Ah i see what you're saying yeah that sounds good 👍

**Comment by emranemran** [comment]
> The current directory structure is like this: image
> 
> So all the fmp4 files are embedded into the `fmp4` directory. I think it's ok, but may be a little confusing for the Transcode API users. Imagine you send the following request:
> 
> ```
> {
>     "input": {
>         "url": "some-video-url"
>     },
>     "storage": {
>         "type": "s3",
>         "endpoint": "https://gateway.storjshare.io",
>         "credentials": {
>             "accessKeyId": "access-key",
>             "secretAccessKey": "secret-key"
>         },
>         "bucket": "my-bucket"
>     },
>     "outputs": {
>         "hls": {
>             "path": "/hls/"
>         },
>         "mp4": {
>             "path": "/mp4/"
>         },
>         "fragmented_mp4": {
>            "path": "/fmp4"
>         }
>     }
> }
> ```
> 
> Now, you could probably expect that your fmp4 files will end it in the `/fmp4` location, but actually they'll end up in `/fmp4/fmp4` location.
> 
> I leave it up to you, but thinking about it again, I'd probably flatten the `fmp4` structure and then for our upload task, we would just add `/fmp4` in the request between task-runner<>catalyst.

Got it, i'll open a ticket for us to make this change. For now only a single customer will likely be using fmp4 output.

**Comment by emranemran** [comment]
> LGTM 👍
> 
> Added one comment again about the directory structure.
> 
> Plus, do we support (want to support) fmp4 in the mediaconvert pipeline?

I think we probably want to in the future to keep both pipelines consistent. For now just livepeer pipeline should be sufficient. @thomshutt wdyt?

**Pull request merged**
